# Initial Install and such

In [ ]:
#Install packages

! pip install transformers datasets
! pip install torch
! pip install tensorflow
! pip install tika
! pip install nltk

from transformers import pipeline
import os
import re
import math
import nltk
from nltk.tokenize import sent_tokenize
from nltk.tokenize import word_tokenize

#nltk.download('all')

Clone GitHub repo with all of the data

In [2]:
!git clone https://github.com/bcarter24/testRepo.git

Cloning into 'testRepo'...
remote: Enumerating objects: 340, done.
remote: Counting objects: 100% (340/340), done.
remote: Compressing objects: 100% (315/315), done.
remote: Total 340 (delta 19), reused 335 (delta 17), pack-reused 0
Receiving objects: 100% (340/340), 6.15 MiB | 23.00 MiB/s, done.
Resolving deltas: 100% (19/19), done.


Move to data directory

In [3]:
%cd /content/testRepo/data/US\ Federal\ Code\ of\ Law/Cleaner_ungrouped/

/content/testRepo/data/US Federal Code of Law/Cleaner_ungrouped


# Read in corpus from US law documents

in the data directory there are several stages of pre-processed data, here I am using the "cleaned-ungrouped" it is just segmented out chapters of the law from federal code, into smaller files so that it can be handled, some of the original was very messy before having it easier to manage in smaller files. it tokenized the input at first, but then I changed to a larger model and used their suggested tokenizer instead.

In [ ]:
directory1 = "./Title18_Chapter44/"
directory2 = "./Title26_Chapter53/"

document_array = []
biggestSentArr = 0
secondLargest = 0
thirdLargest = 0

#enumerate files in the directory into list for directory 1
files_in_direc1 = os.listdir(directory1)
#print(files_in_direc1)
for file in files_in_direc1:
  #check if file is valid
  if os.path.isfile(os.path.join(directory1, file)):
    #open file to read
    file_handle = open(os.path.join(directory1, file), 'r+')
    this_doc = str( file_handle.read())
    file_handle.close()
    #get rid of section part at the beginning but not section references later on
    #this_doc = re.sub(r"§\s478\.", "", this_doc)
    #this_doc = re.sub(r"§", "", this_doc)
    
    #tokenize -- or dont because model changed
    #sentence_array = word_tokenize(this_doc,language='english', preserve_line=True)
    sentence_array = this_doc
    document_array.append(sentence_array)
    print(sentence_array)
    print('length of sentence array is: ' + str(len(sentence_array)) + '\n')
    if len(sentence_array) > biggestSentArr :
      thirdLargest = secondLargest
      secondLargest = biggestSentArr
      biggestSentArr = len(sentence_array)
    #print('current length of Doc array is: ' + str(len(document_array)))

print('\nThere are ' + str(len(document_array)) + ' documents in the document_array')
print('The largest Sentence Array so far has a size of ' + str(biggestSentArr) + ' characters.')
print('The second largest Sentence Array so far has a size of ' + str(secondLargest) + ' characters.')
print('The third largest Sentence Array so far has a size of ' + str(thirdLargest) + ' characters.\n\n\n')


#repeat for directory 2
files_in_direc1 = os.listdir(directory2)
#print(files_in_direc1)
for file in files_in_direc1:
  #check if file is valid
  if os.path.isfile(os.path.join(directory2, file)):
    #open file to read
    file_handle = open(os.path.join(directory2, file), 'r+')
    this_doc = str( file_handle.read())
    file_handle.close()
    #get rid of section part at the beginning but not section references later on
    #this_doc = re.sub(r"§\s478\.", "", this_doc)
    #this_doc = re.sub(r"§", "", this_doc)
    
    #tokenize
    #sentence_array = word_tokenize(this_doc,language='english', preserve_line=True)
    sentence_array = this_doc
    document_array.append(sentence_array)
    print(sentence_array)
    print('length of sentence array is: ' + str(len(sentence_array)) + '\n')
    if len(sentence_array) > biggestSentArr :
      thirdLargest = secondLargest
      secondLargest = biggestSentArr
      biggestSentArr = len(sentence_array)
    #print('current length of Doc array is: ' + str(len(document_array)))

print('\nThere are ' + str(len(document_array)) + ' documents in the document_array')
print('The largest Sentence Array so far has a size of ' + str(biggestSentArr) + ' characters.')
print('The second largest Sentence Array so far has a size of ' + str(secondLargest) + ' characters.')
print('The third largest Sentence Array so far has a size of ' + str(thirdLargest) + ' characters.\n\n\n')

# Get test document (New ATF Memo)

In [5]:
## Credit to Dr. Srivastava for this code section from sample code on github
## located at (https://github.com/biplav-s/course-nl-f22/blob/main/sample-code/l1-wordcloud/FirstLook_ClassSyllabusData.ipynb)

from tika import parser
def convertPdf2TxtWithTika(in_pdf_file, out_text_file):
  print ("INFO: - reading file = " + in_pdf_file)

  raw = parser.from_file(in_pdf_file)
  text = raw['content']
  ## Post-processing explained at: 
  # https://medium.com/@justinboylantoomey/fast-text-extraction-with-python-and-tika-41ac34b0fe61
  # Convert to string
  text = str(text)
  # Ensure text is utf-8 formatted
  safe_text = text.encode('utf-8', errors='ignore')
  # Escape any \ issues
  safe_text = str(safe_text).replace('\\', '\\\\').replace('"', '\\"')
  
  # Write out extracted content
  text_pdf = open(out_text_file, 'w')
  print ("INFO: - writing file = " + out_text_file)
  text_pdf.write(text)
  text_pdf.close()

If you would like to test a particular memo, the place it into the ATF/Testing folder and type in the file name in the code block below

In [28]:
# convert the pdf report to text file
# File needs to be in this folder
directory = "/content/testRepo/data/ATF/Testing/"

# Change the filename here to represent which file you would like to test
fileName = "open_letter_sc_ffls_080820.pdf"

in_file = os.path.join(directory, fileName)
out_file = directory + fileName[:-4] + ".txt"

convertPdf2TxtWithTika(in_file, out_file)

file_handle = open(out_file, 'r+')
clean_up = str( file_handle.read())
file_handle.close()

Memo = re.sub(r"\s+", " ", clean_up)
print(Memo)

INFO: - reading file = /content/testRepo/data/ATF/Testing/open_letter_sc_ffls_080820.pdf
INFO: - writing file = /content/testRepo/data/ATF/Testing/open_letter_sc_ffls_080820.txt
 Microsoft Word - Open Letter South Carolina_Firearms Licensees-080820.docx U.S. Department of Justice Bureau of Alcohol, Tobacco, Firearms and Explosives Washington, DC 20226 August 20, 2008 Open Letter to South Carolina Federal Firearms Licensees The purpose of this letter is to advise you that the recent amendment to Section 16-23-30, Code of Laws of South Carolina, 1976, relating to the unlawful sale of handguns to certain person, so as to prohibit the sale of a handgun to persons under eighteen years of age rather than under twenty-one years of age has no effect on Federal firearms laws and regulations. Your obligations in the conduct of your business remain the same. Change in South Carolina Law: On April 2, 2008, South Carolina Governor Mark Sanford signed South Carolina House Bill 4364 into law. The new

# Test on Model

## Summarize the law sections
Summarize each of the "sections" from a chapter in US Federal Law, each being stored as an element in an array. the array contains the entire corpus of data

In [17]:
# Model used: https://huggingface.co/ahmeddbahaa/mT5_multilingual_XLSum-finetuned-en-cnn

from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

myTokenizer = AutoTokenizer.from_pretrained("ahmeddbahaa/mT5_multilingual_XLSum-finetuned-en-cnn")
myModel = AutoModelForSeq2SeqLM.from_pretrained("ahmeddbahaa/mT5_multilingual_XLSum-finetuned-en-cnn")

classy = pipeline("summarization", model=myModel, tokenizer=myTokenizer, max_length=130, min_length=30, do_sample=False)

Downloading:   0%|          | 0.00/426 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/4.31M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/16.3M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/65.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/854 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.33G [00:00<?, ?B/s]

In [168]:
#takes forever
#summary_array = classy(document_array)

#shorter, just does first index for testing
summary_array = classy(document_array[0:25])

Your max_length is set to 130, but you input_length is only 101. You might consider decreasing max_length manually, e.g. summarizer('...', max_length=50)
Your max_length is set to 130, but you input_length is only 80. You might consider decreasing max_length manually, e.g. summarizer('...', max_length=40)
Your max_length is set to 130, but you input_length is only 33. You might consider decreasing max_length manually, e.g. summarizer('...', max_length=16)


In [169]:
i = 0
length = len(summary_array)
while i < length :
  print(summary_array[i])
  i+=1

{'summary_text': 'Any person who intends to engage in business as a firearms or ammunition importer or manufacturer, or who has not previously been licensed under this part, must file an application for license, ATF Form 7 (Firearms), in duplicate . The application must be executed under the penalties of perjury and thePenalties imposed by 18 U.S.C. 924.'}
{'summary_text': 'The term "includes" and "including" do not exclude other things in the same general class or are otherwise within the scope of the Act . Writers: A person who acquires, holds, or disposes of firearms as curios or relics.'}
{'summary_text': 'Relief from disabilities incurred by indictment . If term of license expires, such importer, manufacturer, dealer must file application for renewal of his license in order to continue operations.'}
{'summary_text': 'Director may issue notice of denial to the applicant . Notice shall set forth matters of fact and law relied upon in determining that the application should be denied

### Example output from the first five indices:

{'summary_text': 'Any person who intends to engage in business as a firearms or ammunition importer or manufacturer, or who has not previously been licensed under this part, must file an application for license, ATF Form 7 (Firearms), in duplicate . The application must be executed under the penalties of perjury and thePenalties imposed by 18 U.S.C. 924.'}

{'summary_text': 'The term "includes" and "including" do not exclude other things in the same general class or are otherwise within the scope of the Act . Writers: A person who acquires, holds, or disposes of firearms as curios or relics.'}

{'summary_text': 'Relief from disabilities incurred by indictment . If term of license expires, such importer, manufacturer, dealer must file application for renewal of his license in order to continue operations.'}

{'summary_text': 'Director may issue notice of denial to the applicant . Notice shall set forth matters of fact and law relied upon in determining that the application should be denied.'}

{'summary_text': 'No firearm barrel shall be imported or brought into the United States unless Director issues a permit authorizing the importation of the firearms . A permit will not be issued for the handgun not generally recognized as suitable for or readily adaptable to sporting purposes.'}

## Summarize the Memo

In [170]:
memoSummary = classy(Memo)
print(memoSummary)

[{'summary_text': 'The new South Carolina law has no effect on Federal firearms laws and regulations . It is unlawful for a person to knowingly sell, offer to sell and deliver, lease, rent, barter, exchange, or transport for sale into this State any handgun to someone under eighteen years of age rather than under twenty-one years - but this shall not apply to the issue of handguns to members of the Armed Forces or the R.O.T.C.'}]


### Summary Text of the Memo for reference:
'The new South Carolina law has no effect on Federal firearms laws and regulations . It is unlawful for a person to knowingly sell, offer to sell and deliver, lease, rent, barter, exchange, or transport for sale into this State any handgun to someone under eighteen years of age rather than under twenty-one years - but this shall not apply to the issue of handguns to members of the Armed Forces or the R.O.T.C.'




This should now reference Title 18, chapter 44, section 922 as the applicable law that is related to this memo (section numbers are different in my corpus since i downloaded them from the ATF site isntead of the federal registry site)

# Compute similarity

In [171]:
import pandas as pd
import sklearn as sk
import numpy as np
import math
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [172]:
# https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html

#for reference this is how you get the text using the keys in the dict
#print(summary_array[1]['summary_text'])

# ATF memo is at 0 index of lower_sumArray
lower_sumArray = []
stringthing = (memoSummary[0].get('summary_text')).lower()
lower_sumArray.append(stringthing)

j = 0
length = len(summary_array)
while j < length :
  thing = (summary_array[j].get('summary_text')).lower()
  lower_sumArray.append(thing)
  j+=1


# taken from scikit learn website
vectorizer = TfidfVectorizer()
X = vectorizer.fit_transform(lower_sumArray)
feature_names = vectorizer.get_feature_names_out()

#print(cosine_similarity (X[0], X[1]))
highestScore = 0
secondHighest = 0
thirdHighest = 0

firstSec = 0
secondSec = 0
thirdSec = 0

for i in range(1, len(lower_sumArray)):
  similarityArr = cosine_similarity (X[0], X[i])
  firstElem = similarityArr[0]
  similarity = round(firstElem[0], 8)
  if similarity > highestScore :
    thirdHighest = secondHighest
    secondHighest = highestScore
    highestScore = similarity
    thirdSec = secondSec
    secondSec = firstSec
    firstSec = i
  print ("ATF Memo with section " + str(i) +" is = "  + str(similarity))

ATF Memo with section 1 is = 0.18678634
ATF Memo with section 2 is = 0.10939218
ATF Memo with section 3 is = 0.05447504
ATF Memo with section 4 is = 0.12108426
ATF Memo with section 5 is = 0.17555682
ATF Memo with section 6 is = 0.1940382
ATF Memo with section 7 is = 0.14402102
ATF Memo with section 8 is = 0.12758658
ATF Memo with section 9 is = 0.14199856
ATF Memo with section 10 is = 0.11062035
ATF Memo with section 11 is = 0.15614477
ATF Memo with section 12 is = 0.24032841
ATF Memo with section 13 is = 0.08396025
ATF Memo with section 14 is = 0.06863849
ATF Memo with section 15 is = 0.1753383
ATF Memo with section 16 is = 0.1562098
ATF Memo with section 17 is = 0.16052546
ATF Memo with section 18 is = 0.16255672
ATF Memo with section 19 is = 0.13976699
ATF Memo with section 20 is = 0.19483858
ATF Memo with section 21 is = 0.14442691
ATF Memo with section 22 is = 0.1030708
ATF Memo with section 23 is = 0.19371188
ATF Memo with section 24 is = 0.14509582
ATF Memo with section 25 is =

### Output for reference

['18', '924', 'acquires', 'act', 'adaptable', 'ammunition', 'an', 'and', 'any', 'applicant', 'application', 'are', 'as', 'atf', 'authorizing', 'barrel', 'be', 'been', 'brought', 'business', 'by', 'class', 'continue', 'curios', 'dealer', 'denial', 'denied', 'determining', 'director', 'disabilities', 'disposes', 'do', 'duplicate', 'engage', 'exclude', 'executed', 'expires', 'fact', 'file', 'firearm', 'firearms', 'for', 'form', 'forth', 'from', 'general', 'generally', 'handgun', 'has', 'his', 'holds', 'if', 'importation', 'imported', 'importer', 'imposed', 'in', 'includes', 'including', 'incurred', 'indictment', 'intends', 'into', 'issue', 'issued', 'issues', 'law', 'license', 'licensed', 'manufacturer', 'matters', 'may', 'must', 'no', 'not', 'notice', 'of', 'operations', 'or', 'order', 'other', 'otherwise', 'part', 'penalties', 'perjury', 'permit', 'person', 'previously', 'purposes', 'readily', 'recognized', 'relics', 'relied', 'relief', 'renewal', 'same', 'scope', 'set', 'shall', 'should', 'sporting', 'states', 'such', 'suitable', 'term', 'that', 'the', 'thepenalties', 'things', 'this', 'to', 'under', 'united', 'unless', 'upon', 'who', 'will', 'within', 'writers']

ATF Memo with section 0 is = [[1.]]
ATF Memo with section 1 is = [[0.25817665]]
ATF Memo with section 2 is = [[0.19790368]]
ATF Memo with section 3 is = [[0.07950102]]
ATF Memo with section 4 is = [[0.17588028]]
ATF Memo with section 5 is = [[0.2415724]]

# The top 3 sections with the highest similarity to the Memo

In [173]:
print("The highest score was section " + str(firstSec) + " with similarity of: " + str(highestScore))
print("The second highest score was section " + str(secondSec) + " with similarity of: " + str(secondHighest))
print("The third highest score was section " + str(thirdSec) + " with similarity of: " + str(thirdHighest))


print("\n\nThis is the original Memo:\n" + Memo + "\n\n\nIt is most similar to the following three sections of the US Federal Code regarding firearm legislation.\nThere is a high probability that the subject matter discussed in this memo could affect the laws discussed in this section, the summaries of these sections are printed below")
print("\n\nHighest Match:\n" + summary_array[firstSec].get('summary_text'))
print("\nSecond highest Match:\n" + summary_array[secondSec].get('summary_text'))
print("\nThird highest Match:\n" + summary_array[thirdSec].get('summary_text'))

The highest score was section 12 with similarity of: 0.24032841
The second highest score was section 6 with similarity of: 0.1940382
The third highest score was section 1 with similarity of: 0.18678634


This is the original Memo:
 Microsoft Word - Open Letter South Carolina_Firearms Licensees-080820.docx U.S. Department of Justice Bureau of Alcohol, Tobacco, Firearms and Explosives Washington, DC 20226 August 20, 2008 Open Letter to South Carolina Federal Firearms Licensees The purpose of this letter is to advise you that the recent amendment to Section 16-23-30, Code of Laws of South Carolina, 1976, relating to the unlawful sale of handguns to certain person, so as to prohibit the sale of a handgun to persons under eighteen years of age rather than under twenty-one years of age has no effect on Federal firearms laws and regulations. Your obligations in the conduct of your business remain the same. Change in South Carolina Law: On April 2, 2008, South Carolina Governor Mark Sanford si